keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configurations
+
+Policy configurations are the interface between `ValidatingAdmissionPolicy`
+authors and the administrators that configure the policies for clusters. The
+configurations allows policy definition authors to define how exactly a
+`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
+schema.
+
+How these configuration CRDs are defined has not yet been decided.
+Alternatives are summarized here and discussed in more detail below:
+
+<<[UNRESOLVED jpbetz, liggitt, ?? ]>>
+Reach consensus on this.
+<<[/UNRESOLVED]>>
+
+| Design Alternative                                | Summary                        |
+| ------------------------------------------------- | ------------------------------ |
+| Duck Typed CRDs                                   | Require CRDs be defined with match field of correct structure |
+| `PolicyConfiguration` kind with config reference  | Match criteria is defined in builtin type which references a config  | 
+| OpenAPIv3 `$ref` in CRDs                          | Match criteria is defined by k8s as a OpenAPIv3 type and CRDs schemas are modified to be allowed to reference it |
+| `/matchRules` subresource                         | Use a subresource to enforce CRDs provide match criteria      |
+| `PolicyConfiguration` kind with config embedded   | Match criteria is defined in a kind, it embeds a config CRD that is freeform |
+| Generate CRDs                                     | Policy defintions contain inline OpenAPIv3 schema for config, CRD is generated with both match and config sections |
+
+###### Alternative: Duck Typed CRDs
+
+This is the alternative shown in the initial examples of this KEP.
+
+Policy authors write a CRD to define how each policy is configured. E.g.:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.rules.example.com
+  annotations:
+    admission.kubernetes.io/is-policy-configuration-definition: "true"
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                config:
+                  type: object
+                    maxReplicas:
+                      type: int
+  scope: Cluster
+  names:
+    kind: ReplicaLimit
+```
+
+The `admission.kubernetes.io/is-policy-configuration-definition` annotation
+means "inject the correct .spec.match during admission and keep it up to date
+with a controller". (Suggested by deads2k). This minimizes version skew if new
+match criteria is added to `spec.match` and also minimizes development effort by
+removing the need to manually declare the fields in CRDs.
+
+The main challenge with this alternative is dealing with mismatches between how
```

Comment:
I think if we ever want to modify this match section it would be extremely difficult with this design.

Reply:
Atomic field problem really messes this alternative up. I've made added a bunch of notes about it and moved it below the other design.